ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ContractsTable.scala

```diff
@@ -0,0 +1,122 @@
+          ORDER BY event_sequential_id DESC
+          LIMIT 1
+       )
+  SELECT contract_id
+    FROM last_contract_key_create -- creation only, as divulged contracts cannot be fetched by key
+   WHERE NOT EXISTS
+         (SELECT 1
+            FROM participant_events, parameters
+           WHERE event_kind = 20 -- consuming exercise
+             AND event_sequential_id <= parameters.ledger_end_sequential_id
+             AND contract_id = last_contract_key_create.contract_id
+         );
+       """
+      .as(contractId("contract_id").singleOpt)
+
+  override final def lookupMaximumLedgerTime(
+      ids: Set[ContractId]
+  )(implicit connection: Connection): Try[Option[Instant]] = {
+    if (ids.isEmpty) {
+      Failure(ContractsTable.emptyContractIds)
+    } else {
+      def lookup(id: ContractId): Option[Option[Instant]] =
+        SQL"""
+  WITH archival_event AS (
+         SELECT participant_events.*
+           FROM participant_events, parameters
+          WHERE contract_id = $id
+            AND event_kind = 20  -- consuming exercise
+            AND event_sequential_id <= parameters.ledger_end_sequential_id
+          LIMIT 1
+       ),
+       create_event AS (
+         SELECT ledger_effective_time
+           FROM participant_events, parameters
+          WHERE contract_id = $id
+            AND event_kind = 10  -- create
+            AND event_sequential_id <= parameters.ledger_end_sequential_id
+          LIMIT 1 -- limit here to guide planner wrt expected number of results
+       ),
+       divulged_contract AS (
+         SELECT ledger_effective_time
+           FROM participant_events, parameters
+          WHERE contract_id = $id
+            AND event_kind = 0 -- divulgence
+            AND event_sequential_id <= parameters.ledger_end_sequential_id
+          ORDER BY event_sequential_id
+            -- prudent engineering: make results more stable by preferring earlier divulgence events
+            -- Results might still change due to pruning.
+          LIMIT 1
+       ),
+       create_and_divulged_contracts AS (
+         (SELECT * FROM create_event)   -- prefer create over divulgance events
+         UNION ALL
+         (SELECT * FROM divulged_contract)
+       )
+  SELECT ledger_effective_time
+    FROM create_and_divulged_contracts
+   WHERE NOT EXISTS (SELECT 1 FROM archival_event)
+   LIMIT 1;
+               """.as(instant("ledger_effective_time").?.singleOpt)
+
+      val foundIds: List[Option[Instant]] = ids.toList
+        .map(lookup)
+        .collect { case Some(found) =>
+          found
+        }
+      val result =
+        if (foundIds.size != ids.size) Failure(ContractsTable.notFound(ids))
+        else {
+          val ledgerTimes = foundIds.collect { case Some(ledgerEffectiveTime) =>
+            ledgerEffectiveTime
+          }
+          val optionalMax: Option[Instant] =
+            if (ledgerTimes.isEmpty) None else Some(ledgerTimes.max)
+          Success(optionalMax)
```

Comment:
Yeah I agree that code sucks :( but unfortunately it is not enough: even if we have found id, it can have no ledgerEffectiveTime, so we cannot know that ledgerTimes will be empty or not.
These are the occasions where I miss scala 2.13 a lot :(

Reply:
Not sure I'm following. `foundIds` is of type `List[Option[Instant]]`. The `Ordered` instance for `Option[Instant]` makes it so that `List(None).max` returns `None` and `List(None, Some(instant)).max` returns `Some(instant)`. This matches the type of `optionalMax: Option[Instant]`.

Reply:
ah I completely missed you point, you are correct, great idea, fixed it. thanks!!

Reply:
kind of lucky that the two orderings aligned perfectly for our use case :)